core/src/main/java/org/apache/gravitino/job/JobManager.java

```diff
@@ -366,15 +370,20 @@ public JobEntity cancelJob(String metalake, String jobId) throws NoSuchJobExcept
                     .withLastModifiedTime(Instant.now())
                     .build())
             .build();
-
-    try {
-      // Update the job entity in the entity store
-      entityStore.put(newJobEntity, true /* overwrite */);
-      return newJobEntity;
-    } catch (IOException e) {
-      throw new RuntimeException(
-          String.format("Failed to update job entity %s to CANCELING status", newJobEntity), e);
-    }
+    return TreeLockUtils.doWithTreeLock(
+        NameIdentifierUtil.ofJob(metalake, jobId),
+        LockType.WRITE,
+        () -> {
+          try {
+            // Update the job entity in the entity store
+            entityStore.put(newJobEntity, true /* overwrite */);
```

Comment:
I wonder when we change the status to `CANCELLED` for a `CANCELLING` job?

Reply:
This will be in another PR.